src/provider.py

```diff
@@ -0,0 +1,185 @@
+
+        Args:
+            event: the event from a related client application needing ACLs
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.kafka_config.zookeeper_connected:
+            logger.debug("cannot update ACLs, ZooKeeper not yet connected")
+            event.defer()
+            return
+
+        provider_relation_config = self.provider_relation_config(event=event)
+        requirer_relation_config = self.requirer_relation_config(event=event)
+
+        self.kafka_auth.load_current_acls()
+
+        self.kafka_auth.update_user_acls(
+            username=provider_relation_config["username"],
+            group=provider_relation_config.get("consumer-group-prefix"),
+            **requirer_relation_config,
+        )
+
+        # non-leader units need cluster_config_changed event to update their super.users
+        self.charm.model.get_relation(PEER).data[self.charm.app].update(
+            {"super-users": self.kafka_config.super_users}
+        )
+
+        event.relation.data[self.charm.app].update(provider_relation_config)
+
+    def _on_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handler for `kafka-client-relation-created` event.
+
+        Adds new relation users to ZooKeeper.
+
+        Args:
+            event: the event from a related client application needing a user
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.ready_to_start:
+            logger.debug("cannot add user, ZooKeeper not yet connected")
+            event.defer()
+            return
+
+        provider_relation_config = self.provider_relation_config(event=event)
+
+        self.kafka_auth.add_user(
+            username=provider_relation_config["username"],
+            password=provider_relation_config["password"],
+        )
+
+        # non-leader units need cluster_config_changed event to update their super.users
+        self.charm.model.get_relation(PEER).data[self.charm.app].update(
+            {provider_relation_config["username"]: provider_relation_config["password"]}
+        )
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handler for `kafka-client-relation-broken` event.
+
+        Removes relation users from ZooKeeper.
+
+        Args:
+            event: the event from a related client application needing a user
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.ready_to_start:
+            logger.debug("cannot remove user, ZooKeeper not yet connected")
+            event.defer()
+            return
+
+        if (
+            event.relation.app != self.charm.app or not self.charm.app.planned_units() == 0
+        ):  # avoid on own charm during teardown
+            provider_relation_config = self.provider_relation_config(event=event)
+
+            self.kafka_auth.load_current_acls()
```

Comment:
Should you handle `subprocess.CalledProcessError` here and in the other call in this file?

Reply:
In this case, I actually want it to fail if it doesn't run correctly.
If there are ACLs, or no ACLs, both return `0`, so if something else goes wrong I'd want to know as soon as possible.